File: 2019/day04.java
public class day04{
  public static void main(String[] args){
    int min = 264793;
    int max = 803935;
    int count = 0;
    int part2 = 0;
    for(int i = min; i <= max; i++){
      int j = i/100000;
      int k = (i%100000)/10000;
      int l = (i%10000)/1000;
      int m = (i%1000)/100;
      int n = (i%100)/10;
      int o = (i%10);
      
      if(j > k){continue;}
      if(k > l){continue;}
      if(l > m){continue;}
      if(m > n){continue;}
      if(n > o){continue;}
      if(!(j == k ||k == l || l == m|| m == n || n == o)){
        continue;
      }
      count++;
      boolean passed = false;
      if(j==k && k==l){
      passed = true;
      if(m == n && n!=o && m!= l){passed = false;}
      if(n ==o && n!=m){passed = false;}
    }
    if(k==l && l ==m){
      passed = true;
      if(n==o && n!=m){passed = false;}
    }
    if(l==m && m == n){
    passed = true;
    if(j==k && k!=m){passed = false;}
    }
    if(m==n && n ==o){
      passed = true;
      if(j ==k && k!=l){passed = false;}
      if(k ==l && j!=k && k!= m){passed = false;}
    }
      if(passed){continue;}
      System.out.println(i);
      part2++;
    }
    System.out.println("Part 1: " + count);
    System.out.println("Part 2: " + part2);
  }
}

File: 2021/python/day04/day04.py
def checkBingo(nums,card):
  soonest = 9999
  for i in range(5):
    hfail = False
    vfail = False
    for j in range(5):
      if card[i][j] not in nums:
        hfail = True
      if card[j][i] not in nums:
        vfail = True
    if(not hfail):
      h = [];
      for j in range(5):
        h.append(nums.index(card[i][j]))
      soonest = min([soonest, max(h)]) 
      
    if(not vfail):
      h = []
      for j in range(5):
        h.append(nums.index(card[j][i]))
      
      soonest = min([soonest, max(h)]) 
      
  return soonest

#nums is shortened to win time
def score(nums, card):
  score = 0
  for i in range(5):
    for j in range(5):
      if card[i][j] not in nums:
        score += card[i][j]
  score = score * nums[-1]
  return score


best = 0
worst = 0
soonest = 9999
lastest = 0
nums = 0
cards = []
with open("2021/day04in.txt") as f:
  nums = f.readline().split(",")
  for i in range(len(nums)):
    nums[i] = int(nums[i])
  input = list(filter(lambda val: val != '\n', f.readlines())) # remove new lines

  while(len(input)):
    card = []
    for i in range(5):
      card.append(input.pop(0));
    cards.append(card)
  for c in cards:
    for i in range(5):
      c[i] = list(filter(lambda val: val != "", c[i].split(" ")))
      for j in range(5):
        c[i][j] = int(c[i][j])
    out = checkBingo(nums,c)
    if out < soonest:
      soonest = out
      best = c
    if out > lastest:
      lastest = out
      worst = c
    
  print("part 1: "  + str(score(nums[0:soonest+1],best)))
  print("part 2: "  + str(score(nums[0:lastest+1],worst)))
    

      
    
  



